Deprecate imprecise DataSize methods, use precise bytes instead

pettyjamesm opened the pull request:
This is an alternative to https://github.com/airlift/units/pull/15 to provide exact value semantics for `DataSize`. Precision is now only lost when using the deprecated construction methods but can be completely preserved by using the new `DataSize.ofBytes` / `DataSize.of(size, Unit)` / `DataSize.succinctBytes` methods.

## Overview
### Before this change:
- DataSize is a pair of (`double size`, `Unit unit`). The size is stored in the given unit, and the unit is preserved unless explicitly converted (ie: `convertToMostSuccinctDataSize()` or `convertTo(Unit)`).
- "Succinct" conversions are defined as choosing the largest unit for which the converted size is greater than or equal to 1.0.
- `DataSize` serializes to JSON using at most 2 decimal places of precision in the current unit which can lead to loss of precision

### Problems with Previous Behavior
- The pairing of "succinct" semantics and JSON encoding can lead to significant rounding error of up to 0.005 of units. Since this occurs per JSON encode / decode round-trip, rounding error can accumulate quickly when aggregating DataSize values.
- Even in-memory DataSize values lose precision by converting the size field between units. Doubles can only represent exact integer values up to 2^53, so `DataSize.toBytes()` for values larger 8,192TB can never be precise even when no conversion between units occurs.

### Investigation of Usages in Presto
- Most `DataSize` instances are constructed via `DataSize.succinctBytes(long)` from an exact byte value. Those usages don't care about preserving the original unit as evidenced by invoking "succinctness" semantics, and are probably unaware of the potential loss of precision.
- Virtually all arithmetic manipulations of `DataSize` do so through `DataSize.toBytes()`. I only found one usage of `getValue() // => double`, and all usages of `getValue(Unit) // => double` take the form of `(long) size.getValue(BYTE)`.

So the existing code seems to want to manipulate these values as if they were exact,  and users probably don't realize how inexact the data sizes reported in query stats can be.

### Changes in this PR
- Changes DataSize to store a `long bytes` field instead of `double size`. Values smaller than 1 byte will no longer be representable. Similarly, values larger than `Long.MAX_VALUE` bytes  (~8 EiB) will also no longer be representable unless we choose to use a `BigInteger` instead. If we do so, we'll have to figure out how to handle the extremely common usage of `DataSize.toBytes()` currently returning a `long`.
- Avoids round-off error in JSON encoding by always serializing the value in bytes. The unit value is discarded during serialization
- Retains the ability to specify human readable values like "1.5GB" in config values by preserving the `DataSize.valueOf(String)` logic and using a different method for JSON deserialization. This method immediately converts to a byte value which is subject to the same precision loss previously present. This is probably OK since these are human specified values, but still worth noting.
- Retains the unit field it was constructed with (except when encoded as JSON), but only only uses it to continue supporting `getUnit()` and `getValue()`.
- Deprecates `succinctDataSize(double, Unit)` and the `DataSize(double, Unit)` constructor for removal. Those methods incur the same conversion to bytes and implied rounding as `DataSize.valueOf(String)`.

### Migration
We can either cut a release with the current deprecations and slowly migrate usage, or we can remove the deprecated methods immediately and change all usages when updating.

Usage of `DataSize` to calculate query "memory seconds" will likely need to use a different representation as they can hit the maximum representable `long` value much more quickly.

aweisberg commented:
I think it's fine to remove deprecated methods and make updating a requirement of updating airlift as a dependency, but it's also OK to add the new ones first.

How much of this loss of precision occurs just due to the loss of significant digits in JSON? If that's fixed is it good enough?

It seems pretty reasonable to separate precise  memory values from approximate and very large. The JSON changes seem like a big improvement.

rschlussel commented:
Generally makes sense to me and seems very reasonable to know whether you're dealing with a precise or approximate data size.  Do you know if there are any other usages aside from memory seconds that would exceed MAX_LONG?

pettyjamesm commented:
@rschlussel that’s easiest way I’m  currently aware of for a value to overflow in a “typical” use case at this point. Aggregated values in query stats could conceivably hit the limit too, but they’d  pretty much have to hit the limit on memory seconds first. I’m hoping that if anyone is running such an impressive deployment, they’ll speak up- at which point we could start talking about using BigInteger instead.

pettyjamesm commented:
@martint I've addressed the stylistic comments on the PR and found a way to remove the `DataSize.fromBytesValueString(String)` method. Instead, `DataSize.valueOf(String)` will now parse the numeric value as either `double` or `long` based on whether it contains a `'.'` character.

Since the proposal right now is to have a deprecation cycle, I'll take a look and come up with a proposal for deprecating some additional public instance methods on the class. If we change our mind about deprecations, those methods would also be removed immediately. Stay tuned.

pettyjamesm commented:
I've added a new commit with the following public method groups deprecated:

### `DataSize#getValue`
These methods expose floating point semantics which encourage loss of precision
- `DataSize#getValue()` - units basically never known anyway, very few usages
in presto and all easily replaced with `DataSize#toBytes()`
- `DataSize#getValue(Unit)` - only one usage found in presto, also easily replaced
with `DataSize#toBytes()`

### `DataSize#convert`
The name for these methods is misleading since altering the unit no longer performs a conversion of the underlying value. New methods `DataSize#to(Unit)` and `DataSize#succinct()` are direct replacements.
- `DataSize#convertTo(Unit)` - use `DataSize#to(Unit)` instead
- `DataSize#convertToMostSuccinctDataSize()` - use `DataSize#succinct()`
  instead

### `DataSize#round`
This method relies on floating point arithmetic semantics which are deprecated elsewhere. Since possible alternatives would greatly increase the complexity of this class and no usages currently exist in presto, I propose deprecating it without replacement.

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -23,55 +23,103 @@
+     * @throws IllegalArgumentException If the provided size would overflow a long value when converting to bytes
+     */
+    public static DataSize of(long size, Unit unit)
+        throws IllegalArgumentException
```
Nit: continuation indent is 8 spaces

pettyjamesm replied:
Changed

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -23,55 +23,103 @@
+            return new DataSize(size, unit);
+        }
+        try {
+            return new DataSize(Math.multiplyExact(size, unit.getFactor()), unit);
```
Static import `multiplyExact`

electrum replied:
Check for negative size before multiplying

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -23,55 +23,103 @@
+            return new DataSize(Math.multiplyExact(size, unit.getFactor()), unit);
+        }
+        catch (ArithmeticException e) {
+            throw new IllegalArgumentException("size is too large to be represented in bytes");
```
Let's include the bad value here
```java
throw new IllegalArgumentException(format("size is too large to be represented in bytes: %s%s", size, unit.getUnitString());

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -131,27 +194,37 @@ public static DataSize valueOf(String size)
+            // Strings without decimals can avoid precision loss by parsing as long
+            return DataSize.of(Long.parseLong(number), unit);
+        }
+        else {
```
Nit: else is redundant

pettyjamesm replied:
Changed.

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -131,27 +194,37 @@ public static DataSize valueOf(String size)
+        requireNonNull(unit, "unit is null");
+        double rounded = floor((size / (1.0d / unit.getFactor())) + 0.5d);
+        checkArgument(rounded <= Long.MAX_VALUE,
+                "size is too large to be represented in requested unit as a long");
```
We might want to include the size+unit value here (same as in the `of()` factory)

pettyjamesm replied:
Added.

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -32,28 +32,53 @@
-        assertEquals(succinctBytes(123), new DataSize(123, BYTE));
-        assertEquals(succinctBytes((long) (5.5 * 1024)), new DataSize(5.5, KILOBYTE));
-        assertEquals(succinctBytes(5 * 1024 * 1024), new DataSize(5, MEGABYTE));
+        DataSize oneHundred = DataSize.ofBytes(123);
```
This variable name is misleading. I think inlining here is clearer:
```java
assertEquals(succinctBytes(123), DataSize.ofBytes(123));
assertEquals(succinctBytes(123).getUnit(), BYTE);

pettyjamesm replied:
Changed.

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -32,28 +32,53 @@
+    @Test
+    public void testDeprecatedSuccinctFactories()
+    {
+        assertEquals(succinctDataSize(123, BYTE), DataSize.of(123, BYTE));
```
Should this be `DataSize.ofBytes(123)`?

pettyjamesm replied:
Changed

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -65,23 +90,45 @@ public void testConvertToMostSuccinctDataSize(DataSize.Unit unit, DataSize.Unit
                 .addLesserGroup(group(0))
                 .addGreaterGroup(group(1))
                 .addGreaterGroup(group(123352))
-                .addGreaterGroup(group(Long.MAX_VALUE))
```
Why remove this? It seems like we should have some larger values here (at least larger than MAX_INT)

pettyjamesm replied:
Probably to make an earlier iteration work, I'll add it back since it works again with this version.

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -120,7 +167,16 @@ public void testValueOf(String string, double expectedValue, DataSize.Unit expec
+    {
+        //  If parsed as double, this results in only 2^53 not (2^53) + 1
+        DataSize tooLargeAsDouble = DataSize.ofBytes((1L << 53) + 1);
+        assertEquals(DataSize.valueOf(tooLargeAsDouble.toBytesValueString()), tooLargeAsDouble, "should parse as long and not double");
```
Can we make this
```java
Long.toString(tooLargeAsDouble.toBytes()) + "B"
```
So that it's the same as the one below?

pettyjamesm replied:
Changed.

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -32,28 +32,53 @@
+    public void testToBytesValueString()
+    {
+        DataSize oneByte = DataSize.ofBytes(1);
+        assertEquals(oneByte.toString(), oneByte.toBytesValueString(), "exact values match toString()");
```
Let's verify the actual output as well
```java
assertEquals(oneByte.toString(), "1B");

pettyjamesm replied:
Changed.

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -32,28 +32,53 @@
+        DataSize oneByte = DataSize.ofBytes(1);
+        assertEquals(oneByte.toString(), oneByte.toBytesValueString(), "exact values match toString()");
+
+        for (DataSize.Unit unit : DataSize.Unit.values()) {
```
Add a few tests with fixed expected output, so it's easy to see and verify the expected behavior:
```java
assertEquals(DataSize.of(8, MEGABYTE).toBytesValueString(), "8388608B");
```

pettyjamesm replied:
Added

electrum commented on src/test/java/io/airlift/units/TestDataSize.java:
```diff
@@ -244,12 +316,19 @@ public void testRoundToTooLarge3()
+
+        for (DataSize.Unit unit : DataSize.Unit.values()) {
+            assertJsonRoundTrip(DataSize.ofBytes(1L).convertTo(unit));
+            assertJsonRoundTrip(DataSize.of(1, unit));
```
Also test values other than `1`

pettyjamesm replied:
Added

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -115,8 +115,9 @@ public long toBytes()
 
     /**
-     * Prefer {@link DataSize#toBytes()} since it avoids conversion to a floating point representation
+     * @deprecated Use {@link DataSize#toBytes()} instead to avoid floating point precision semantics
```
I'm not sure deprecating this make sense, as it basically makes the stored unit useless. Why have `to(Unit)` at all in that case (since you can't get the fractional value in that unit)?

martint replied:
The unit is useful for printing (i.e., `toString`)

pettyjamesm replied:
Indeed. I marked all methods that expose the fractional semantics as deprecated instead of outright removing them. This method clearly falls in that category. The preservation of unit (except when serialized and deserialized as JSON) is kept since printing config values in the same format  as they were specified has some value, but accessing the fractional value should be discouraged except in the toString() context.

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -147,23 +158,47 @@ public long roundTo(Unit unit)
                 unitToUse = unitToTest;
-            }
-            else {
+            } else {
```
Nit: formatting -- else goes on separate line

pettyjamesm replied:
Changed.

electrum commented on src/main/java/io/airlift/units/DataSize.java:
```diff
@@ -147,23 +158,47 @@ public long roundTo(Unit unit)
-    }
-
-    public DataSize convertToMostSuccinctDataSize()
+    public Unit succinctUnit()
```
This should probably be private

pettyjamesm replied:
Changed.

pettyjamesm commented:
Addressed PR comments in a new commit instead of trying to separate them out into the base commit vs extended deprecations commit. Will rebase the final agreed upon changes into a single commit once it's in an agreed upon state.

pettyjamesm commented:
Squashed the commits together to prepare for merge.
